File: internal/core/spec.go
package core

import (
	"fmt"
	"strings"
)

const (
	SpecVersion = "1.0"
	AgentMD     = "agent.md"
	AgentHTML   = "agent.html"
	LLMSTxt     = "llms.txt"

	InfoKey      = "AgentReadability"
	InfoValue    = "agentic-pdf-spec/" + SpecVersion
	CanonicalKey = "CanonicalSource"
)

// MarkerKeywords are written to the PDF Keywords field so corpus-wide
// indexing can discover agentic files.
var MarkerKeywords = []string{"agent-readable", "agentic-pdf", "llm-readable"}

type Frontmatter struct {
	Title       string
	Description string
	DocVersion  string
	LastUpdated string
	Canonical   string
	Generator   string
	SourcePages int
}

func (fm Frontmatter) Render() string {
	var b strings.Builder
	b.WriteString("---\n")
	fmt.Fprintf(&b, "title: %q\n", fm.Title)
	fmt.Fprintf(&b, "description: %q\n", fm.Description)
	fmt.Fprintf(&b, "doc_version: %q\n", fm.DocVersion)
	fmt.Fprintf(&b, "last_updated: %q\n", fm.LastUpdated)
	if fm.Canonical != "" {
		fmt.Fprintf(&b, "canonical: %q\n", fm.Canonical)
	}
	if fm.SourcePages > 0 {
		fmt.Fprintf(&b, "source_pages: %d\n", fm.SourcePages)
	}
	fmt.Fprintf(&b, "generator: %q\n", fm.Generator)
	b.WriteString("---\n")
	return b.String()
}

// ParseFrontmatter splits a leading YAML frontmatter block from the body.
func ParseFrontmatter(md string) (map[string]string, string) {
	fm := map[string]string{}
	s := strings.TrimPrefix(md, "---\n")
	if s == md {
		return fm, md
	}
	end := strings.Index(s, "\n---")
	if end < 0 {
		return fm, md
	}
	for _, line := range strings.Split(s[:end], "\n") {
		k, v, ok := strings.Cut(line, ":")
		if !ok {
			continue
		}
		v = strings.TrimSpace(v)
		v = strings.Trim(v, `"`)
		fm[strings.TrimSpace(k)] = v
	}
	body := s[end+4:]
	body = strings.TrimPrefix(body, "\n")
	return fm, body
}

File: src/core/spec.ts
export const SPEC_VERSION = "1.0";

export const AGENT_MD = "agent.md";
export const AGENT_HTML = "agent.html";
export const LLMS_TXT = "llms.txt";

export const INFO_KEY = "AgentReadability";
export const INFO_VALUE = `agentic-pdf-spec/${SPEC_VERSION}`;

export const MARKER_KEYWORD = "agent-readable";

export interface AgentFrontmatter {
  title: string;
  description: string;
  doc_version: string;
  last_updated: string;
  canonical?: string;
  generator: string;
  source_pages?: number;
}

export function renderFrontmatter(fm: AgentFrontmatter): string {
  const lines = ["---"];
  lines.push(`title: ${JSON.stringify(fm.title)}`);
  lines.push(`description: ${JSON.stringify(fm.description)}`);
  lines.push(`doc_version: ${JSON.stringify(fm.doc_version)}`);
  lines.push(`last_updated: ${JSON.stringify(fm.last_updated)}`);
  if (fm.canonical) lines.push(`canonical: ${JSON.stringify(fm.canonical)}`);
  if (fm.source_pages != null) lines.push(`source_pages: ${fm.source_pages}`);
  lines.push(`generator: ${JSON.stringify(fm.generator)}`);
  lines.push("---");
  return lines.join("\n");
}

export function parseFrontmatter(md: string): { frontmatter: Record<string, string>; body: string } {
  const m = /^---\r?\n([\s\S]*?)\r?\n---\r?\n?/.exec(md);
  if (!m) return { frontmatter: {}, body: md };
  const frontmatter: Record<string, string> = {};
  for (const line of m[1].split(/\r?\n/)) {
    const kv = /^([A-Za-z_][\w-]*):\s*(.*)$/.exec(line);
    if (!kv) continue;
    let value = kv[2].trim();
    const q = /^"(.*)"(\s*#.*)?$/.exec(value) ?? /^'(.*)'(\s*#.*)?$/.exec(value);
    if (q) value = q[1];
    frontmatter[kv[1]] = value;
  }
  return { frontmatter, body: md.slice(m[0].length) };
}
